fbpic/print_utils.py

```diff
@@ -0,0 +1,234 @@
+                message += '\nBoosted frame gamma: %d' \
+                    %sim.comm.gamma_boost
+                if sim.use_galilean:
+                    message += '\nGalilean frame: Yes'
+                else:
+                    message += '\nGalilean frame: No'
+            else:
+                message += '\nBoosted frame: False'
+        message += '\n'
+        print( message )
+    if level == 2:
+        # Sync MPI processes before printing MPI GPU selection
+        sim.comm.mpi_comm.barrier()
+        time.sleep(0.1)
+        if sim.use_cuda:
+            print_current_gpu( MPI )
+            sim.comm.mpi_comm.barrier()
+            time.sleep(0.1)
+            if sim.comm.rank == 0:
+                print('')
+
+def progression_bar( i, Ntot, avg_time_per_step, prev_time,
+                     n_avg=20, Nbars=35, char=progress_char):
+    """
+    Shows a progression bar with Nbars and the estimated
+    remaining simulation time.
+    """
+    # Estimate average time per step
+    curr_time = time.time()
+    time_per_step = curr_time - prev_time
+    avg_time_per_step += (time_per_step - avg_time_per_step)/n_avg
+    if i <= 2:
+        # Ignores first step in time estimation (compilation time)
+        avg_time_per_step = time_per_step
+    # Print progress bar
+    if i == 0:
+        # Let the user know that the first step is much longer
+        sys.stdout.write('\r' + \
+            'Just-In-Time compilation (up to one minute) ...')
+        sys.stdout.flush()
+    else:
+        # Print the progression bar
+        nbars = int( (i+1)*1./Ntot*Nbars )
+        sys.stdout.write('\r' + nbars*char )
+        sys.stdout.write((Nbars-nbars)*' ')
+        sys.stdout.write(' %d/%d' %(i+1, Ntot))
+        if i < n_avg:
+            # Time estimation is only printed after n_avg timesteps
+            sys.stdout.write(', calc. ETA...')
+        else:
+            # Estimated time in seconds until it will finish
+            eta = avg_time_per_step*(Ntot-i)
+            # Conversion to H:M:S
+            m, s = divmod(eta, 60)
+            h, m = divmod(m, 60)
+            sys.stdout.write(', %d:%02d:%02d left' % (h, m, s))
+        # Time taken by the last step
+        sys.stdout.write(', %d ms/step' %(time_per_step*1.e3))
+        sys.stdout.flush()
+    # Clear line
+    sys.stdout.write('\033[K')
+
+    return avg_time_per_step, curr_time
+
+def print_runtime_summary( N, duration ):
+    """
+    Print a summary about the total runtime of the simulation.
+
+    Parameters
+    ----------
+    N: int
+        The total number of iterations performed by the step loop
+
+    duration: float, seconds
+        The total time taken by the step loop
+    """
+    avg_tps = (duration / N)*1.e3
+    m, s = divmod(duration, 60)
+    h, m = divmod(m, 60)
+    print('\nTime taken (with compilation): %d:%02d:%02d' %(h, m, s))
+    print('Average time per iteration ' \
+          '(with compilation): %d ms\n' %(avg_tps))
+
```

Comment:
I wonder if we should remove the `print_gpu_meminfo` functions. It seems that in most cases `nvidia-smi` would do the job, no?

Reply:
You are right, but I would still keep this function for the moment. It might be useful at some point for debugging as we can use it to quickly track the memory consumption at different places in the loop during execution.